virttest/qemu_vm.py

```diff
@@ -4015,57 +4014,49 @@ def _go_down(session, timeout):
             except Exception:
                 return True
 
-        def wait_for_go_down(session, timeout):
-            """ Wait for VM go down
-            :param session: VM session object
-            :param timeout: timeout wait for VM go down
-            """
-            try:
-                if not utils_misc.wait_for(
-                        lambda: _go_down(session, timeout),
-                        timeout=timeout):
-                    raise virt_vm.VMRebootError("Guest refuses to go down")
+        def _shell_reboot(session, timeout):
+            if not session:
+                if not serial:
+                    session = self.wait_for_login(nic_index=nic_index,
+                                                  timeout=timeout)
                 else:
-                    logging.info("VM go down, wait for it boot up")
-            finally:
-                session.close()
-
-        def shell_reboot(session, timeout):
-            """
-            Reboot guest OS via shell command
-
-            :param session: VM session object
-            :param timeout: timeout wait for VM go down
-            """
+                    session = self.wait_for_serial_login(timeout=timeout)
             reboot_cmd = self.params.get("reboot_command")
-            step_info = "Send %s command, wait for VM rebooting" % reboot_cmd
-            error_context.context(step_info)
+            logging.debug("Send command: %s" % reboot_cmd)
             session.sendline(reboot_cmd)
-            start_time = time.time()
-            wait_for_go_down(session, timeout=timeout)
-            return int(time.time() - start_time)
 
         error_context.base_context("rebooting '%s'" % self.name, logging.info)
         error_context.context("before reboot")
         error_context.context()
-        shutdown_dur = 0
+
+        start_time = time.time()
         if method == "shell":
-            # ensure VM bootup really
-            if not session:
-                if not serial:
-                    session = self.wait_for_login(nic_index=nic_index,
-                                                  timeout=timeout)
-                else:
-                    session = self.wait_for_serial_login(timeout=timeout)
-            # send reboot command via session
-            shutdown_dur = shell_reboot(session, timeout)
+            _reboot = partial(_shell_reboot, session, timeout)
+            _check_go_down = partial(_go_down, session, timeout)
         elif method == "system_reset":
-            self.system_reset()
+            _reboot = self.system_reset
+            _check_go_down = partial(bool, True)
         else:
             raise virt_vm.VMRebootError("Unknown reboot method: %s" % method)
+
+        if isinstance(self.monitor, qemu_monitor.QMPMonitor):
+            _check_go_down = partial(self.monitor.get_event, "RESET")
```

Comment:
The `_check_go_down` function will be called at line 4049, and the next line will clear `RESET` events, so I think it should be ok.